### ISSUE TITLE
AuthorizationHeaderMalformed from GetBucketLocation()

### ISSUE THREAD

**Issue by lsong0905** [aws/aws-sdk-cpp#1339]
Hello,

I am trying to understand what is happening underneath the hood when trying to call GetBucketLocation().

The version of sdk I am using is 1.7.29. 
My application attempts to get the bucket's region before sending the request. In my test case, the bucket that I am interested in exists in us-west-2. The below is the step I run my test. 
1. I create a ClientConfiguration and set "endpointOverride" to "s3-us-west-2.amazonaws.com", which is passed from a user.
2. Call GetBucketLocation() with the bucket name, and it failed with the error message "AuthorizationHeaderMalformed Message: The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'us-west-2'" 

Here is my question. 
1. It looks like AWSAuthV4Signer always goes to us-east-1 to sign regardless of the bucket region. Is this expected? and what is the reason behind this?

```
AWSAuthV4Signer [140734514587392] Final String to sign: AWS4-HMAC-SHA256
1643 20200311T162000Z
1644 20200311/us-east-1/s3/aws4_request
```

2. In my understanding (Please correct me if I am wrong), when we override the endpoint with a valid url "s3-us-west-2.amazonaws.com" in ClientConfiguration, GetBucketLocation() will ask us-west-2 to find the region of the bucket. But what does the error message "The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'us-west-2'" truly mean? Does this mean that when the the user calls GetBucketLocation(), is the endpoint always supposed to be set to "s3.us-east-1.amazonaws.com" or just "s3.amazonaws.com" (which will route to us-east-1)? because I do not get the error if the endpoint is set to us-east-1. More interestingly, where does "us-east-1" in the message come from?

**Comment by KaibaLopez** [comment]
Hi @lsong0905 ,
hard to say without taking a look at the code you're running, but my first guess would be that you're not specifying the region for the s3 client.
So a few things I'll ask to be able to help:
* Try specifying the region on your s3 client.
```
Aws::Client::ClientConfiguration clientConfig();
clientConfig.region="us-west-2";
Aws::S3::S3Client clS3Client(clientConfig);
```
* If that doesn't work, can you provide sample code that would reproduce the error? 
* Can you activate trace logs, sanitize them for personal info, and post them?
```
 options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Trace;
 Aws::InitAPI(options);
```

**Comment by lsong0905** [comment]
@KaibaLopez No. I do not specify the region. The purpose of this application is to find the region of the bucket prior to the request. 

To be more specific, we create a dummy client to find the region of the bucket. Here is a partial code from the application. 

```
...
Aws::Client::ClientConfiguration client_config;
client_config.endpointOverride = Aws::String(user_url);
Aws::S3::S3Client *cli = Aws::S3::S3Client(...,client_config,...);
...
Aws::S3::Model::GetBucketLocationRequest request;
auto outcome = cli->GetBucketLocation(request);
...
```

Then I get the error from GetBucketLocation(request). The below is partial error messages after I sanitized.

```
getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 AWSAuthV4Signer [140737353987904] Canonical Header String: content-type:application/xml
host:s3.us-west-2.amazonaws.com
x-amz-api-version:2006-03-01
x-amz-content-sha256:UNSIGNED-PAYLOAD
x-amz-date:20200311T185334Z


getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 AWSAuthV4Signer [140737353987904] Signed Headers value:content-type;host;x-amz-api-version;x-amz-content-sha256;x-amz-date

 getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 AWSAuthV4Signer [140737353987904] Canonical Request String: GET
/USER_BUCKET
location=
content-type:application/xml
host:s3.us-west-2.amazonaws.com
x-amz-api-version:2006-03-01
x-amz-content-sha256:UNSIGNED-PAYLOAD
x-amz-date:20200311T185334Z

content-type;host;x-amz-api-version;x-amz-content-sha256;x-amz-date
UNSIGNED-PAYLOAD

 getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 AWSAuthV4Signer [140737353987904] Final String to sign: AWS4-HMAC-SHA256
20200311/us-east-1/s3/aws4_request

 getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 AWSAuthV4Signer [140737353987904] Final computed signing hash: 

getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 AWSClient [140737353987904] Request Successfully signed
getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:34 CurlHttpClient [140737353987904] Making request to https://s3.us-west-2.amazonaws.com/USER_BUCKET?location

getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:34 CurlHttpClient [140737353987904] content-type: application/xml

getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:34 CurlHttpClient [140737353987904] host: s3.us-west-2.amazonaws.com

getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:34 CurlHttpClient [140737353987904] user-agent: aws-sdk-cpp/1.7.29 Linux/4.4.114-94.11-default x86_64 GCC/4.8.5

getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:34 CurlHttpClient [140737353987904] x-amz-api-version: 2006-03-01

getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:34 CurlHttpClient [140737353987904] x-amz-content-sha256: UNSIGNED-PAYLOAD

getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:34 CurlHandleContainer [140737353987904] Attempting to acquire curl connection.
...
getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:35 CurlHttpClient [140737353987904] HTTP/1.1 400 Bad Request
getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:35 CurlHttpClient [140737353987904] Returned http response code 400
...
getbucketlocation[31072]:Debug: [DEBUG] 2020-03-11 18:53:35 AWSClient [140737353987904] Request returned error. Attempting to generate appropriate error codes from response

getbucketlocation[31072]:Debug: [TRACE] 2020-03-11 18:53:35 AWSErrorMarshaller [140737353987904] Error response is <?xml version="1.0"?>
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <Code>AuthorizationHeaderMalformed</Code>
    <Message>The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'us-west-2'</Message>
    <Region>us-west-2</Region>
    <RequestId></RequestId>
</Error>
```
Please let me know if you need more information.

**Comment by lsong0905** [comment]
@KaibaLopez Any update on this?

**Comment by KaibaLopez** [comment]
Ok, so I've tested this and you're right, this should not happen, but we do have a fix coming up that should take care of this.
Basically the way we are currently managing regions is not the best and we are changing it for cases like this one, where a sort of "global " region is the best answer.
For now though, as a work around because the fix might still take some time, set the configuration's region to "us-west-1" and that will work as a global region and the call will succeed regardless of the bucket's region.
Try it out and let me know if it worked for you too.

**Comment by lsong0905** [comment]
@KaibaLopez Thanks for the reply. 
I've tried to set the region to "us-west-1" prior to getbucketlocation(), and still failed. Same exception code with "us-west-1" instead of "us-east-1" this time. 

Added this line.
`client_config.region = "us-west-1";`


```
581 <Error>
582     <Code>AuthorizationHeaderMalformed</Code>
583     <Message>The authorization header is malformed; the region 'us-west-1' is wrong; expecting 'us-west-2'</Message>
584     <Region>us-west-2</Region>
585     <RequestId></RequestId>
586     <HostId></HostId>
587 </Error>
```

So, how should I as a end user interpret this error message? Is it saying that the default region (us-west-1) in this case conflicts with the endpoint (s3.us-west-2.amazonaws.com)? And how does "global" region work?

**Comment by wps132230** [comment]
Hi there, to get bucket location, you don't need to override the endpoint.

For the original question, the endpoint is `s3-us-west-2.amazonaws.com`, and the signer region is `us-east-1` by default. Then you will get "The authorization header is malformed" error.

To avoid this error you should have both endpoint and signer in the same region:
1. If you want to override endpoint: `s3-us-west-2.amazonaws.com`, then you should set signer region to be `us-west-2` by `clientconfiguration.region = Region::US_WEST_2;`.
2. Or you don't need to override endpoint. The default region is `us-east-1`, then the default endpoint will be `s3.amazonaws.com`, even though it's a different region from where your bucket located, you can get the location successfully. And you will not get the signature error.

**Comment by lsong0905** [comment]
@wps132230 Thanks for the reply. Yes. After I matched the endpoint and the region, it seems working ok. So, is this a bug as @KaibaLopez mentioned? Or is it expected to the users to match the endpoint and region?

**Comment by KaibaLopez** [comment]
Hi @lsong0905 ,
So, I answered too fast and I did not understand the root cause of the error when I first responded sorry about the confusion. 
It is not so much a bug but a feature we are missing I guess, but it is on the way of being added that part still applies. wps explanation is correct, we currently expect you to match the endpoints, but in the future this will be done for you so you wouldn't have to worry about it.

**Comment by lsong0905** [comment]
@KaibaLopez Okay. Thank you for the help!

**Comment by singku-china** [comment]
@lsong0905  We have published [SDK 1.8 preview build](https://github.com/aws/aws-sdk-cpp/tree/version1.8) which include `aws global` pseudo  region feature that allows SDK to identify the correct region or make cross region request call for you. Check out the [blogpost](https://aws.amazon.com/blogs/developer/aws-sdk-for-c-version-1-8-developer-preview/?utm_source=feedburner&utm_medium=feed&utm_campaign=Feed%3A+AwsDeveloperBlog+%28AWS+Developer+Blog%29) and search for `aws global`, or take a look at our [Github wiki page](https://github.com/aws/aws-sdk-cpp/wiki/What%E2%80%99s-New-in-AWS-SDK-for-CPP-Version-1.8).
